### ISSUE TITLE
Test producing different result after ef38880

### ISSUE THREAD

**Issue by dyung** [llvm/llvm-project#62242]
We have an internal test which started failing recently which I bisected to commit ef38880. Consider the following code:
```c++
#include <stdio.h>
void c(void *d, unsigned e) {
  unsigned char *bytes = (unsigned char *)d;
  char f[3];
  for (unsigned g = 0; g != e; ++g)
    sprintf(f, "%02x", bytes[g]);
  printf(f);
}
typedef int i;
typedef char __attribute__((ext_vector_type(32))) j;
typedef short __attribute__((ext_vector_type(16))) k;
typedef union {
  j a;
  k b;
} l;
typedef struct {
  k a;
} m;
typedef i __attribute__((ext_vector_type(4))) n;
typedef i __attribute__((ext_vector_type(8))) o;
typedef short __attribute__((ext_vector_type(8))) p;
j aa;
n q{3, 60877, 6, 70189285}, r{2, 308336, 5, -17194088},
    s{5, 0, 80290862, 31620660}, ab = r ^ s,
                                 ac{40069418, 5079601, 1, 805389106};
p t;
int main() {
  n u{3058, 7, 4, 4005636};
  q &= ab;
  n v(q | ac);
  o w = __builtin_shufflevector(u, v, 6, 7, 3, 1, 7, 3, 3, 2);
  p x(__builtin_convertvector(w, p));
  k y = __builtin_shufflevector(x, t, 12, 5, 12, 7, 5, 12, 0, 7, 13, 9, 0, 12,
                                8, 12, 9, 8);
  l z;
  m ad;
  ad.a = y;
  z.b = ad.a;
  aa = z.a;
  char ae = z.a[8];
  c(&ae, sizeof(z));
}
```
When compiled with optimizations and AVX2 (`-O2 -mavx2`) the program used to print out "04" but after ef38880 now prints "b6" instead.

Link showing the output of trunk for the above program vs. LLVM 16: https://godbolt.org/z/qGozc8aWc

**Comment by zhuhan0** [comment]
Reduced to LLVM IR:
```
@q = global <4 x i32> <i32 3, i32 60877, i32 6, i32 70189285>, align 16
@ab = global <4 x i32> <i32 7, i32 308336, i32 80290859, i32 -14950996>, align 16
@ac = global <4 x i32> <i32 40069418, i32 5079601, i32 1, i32 805389106>, align 16
@t = global <8 x i16> zeroinitializer, align 16

; Function Attrs: nofree nounwind
declare i32 @printf(ptr nocapture noundef readonly, ...)

; Function Attrs: mustprogress nofree norecurse nounwind uwtable
define <16 x i16> @main() {
entry:
  %f.i = alloca [3 x i8], align 1
  %0 = load <4 x i32>, ptr @ab, align 16
  %1 = load <4 x i32>, ptr @q, align 16
  %and = and <4 x i32> %1, %0
  store <4 x i32> %and, ptr @q, align 16
  %2 = load <4 x i32>, ptr @ac, align 16
  %or = or <4 x i32> %2, %and
  %shuffle = shufflevector <4 x i32> <i32 poison, i32 7, i32 4, i32 4005636>, <4 x i32> %or, <8 x i32> <i32 6, i32 7, i32 3, i32 1, i32 7, i32 3, i32 3, i32 2>
  %conv = trunc <8 x i32> %shuffle to <8 x i16>
  %3 = load <8 x i16>, ptr @t, align 16
  %shuffle1 = shufflevector <8 x i16> %conv, <8 x i16> %3, <16 x i32> <i32 12, i32 5, i32 12, i32 7, i32 5, i32 12, i32 0, i32 7, i32 13, i32 9, i32 0, i32 12, i32 8, i32 12, i32 9, i32 8>
  ret <16 x i16> %shuffle1
}
```
While lowering this, there's difference between the main branch and with ef38880 reverted.

Without ef38880:
```
                                    t100: v4i32 = extract_subvector t81, Constant:i64<0>
                                          t86: i64 = X86ISD::WrapperRIP TargetGlobalAddress:i64<ptr @ac> 0 [TF=5]
                                        t87: i64,ch = load<(load (s64) from got)> t0, t86, undef:i64
                                      t43: v4i32,ch = load<(dereferenceable load (s128) from @ac)> t0, t87, undef:i64
                                    t12: v4i32 = or t43, t7
                                  t138: v4i32 = X86ISD::UNPCKH t100, t12
```
Main branch:
```
                                    t12: v4i32 = or t43, t7
                                        t95: i64 = X86ISD::Wrapper TargetConstantPool:i64<<8 x i32> <i32 undef, i32 undef, i32 4, i32 4005636, i32 undef, i32 undef, i32 undef, i32 undef>> 0
                                      t81: v8i32,ch = load<(load (s256) from constant-pool)> t0, t95, undef:i64
                                    t100: v4i32 = extract_subvector t81, Constant:i64<0>
                                  t138: v4i32 = X86ISD::UNPCKH t12, t100
```
Somehow the order of t12 and t100 are swapped.

**Comment by zhuhan0** [comment]
Looks like there's a bug in the permute mask calculation.

**Comment by zhuhan0** [comment]
Fix up: https://reviews.llvm.org/D148843

**Comment by zhuhan0** [comment]
Fixed by 075202d.
